### PR TITLE
🏃 Adds machine filters and FilterableMachineCollection

### PR DESCRIPTION
**Pull request by chuckha**
I'm pulling apart #2334 to help @detiber focus on finishing out implementation and get a smaller PR down the line. We discussed this work in slack.

Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR adds FilterableMachineCollection, which is basically a rename of a set data structure containing machines, but avoiding the name MachineSet.

This also adds the filters to be used in following PRs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #2242

### CODE REVIEW

File: controlplane/kubeadm/internal/machine_set.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/*
```
Comment (detiber) [inline controlplane/kubeadm/internal/machine_set.go]: :grimacing: It looks like I had forgotten to rename this file from `machine_set.go`, might be good to do here.

Comment (chuckha) [reply controlplane/kubeadm/internal/machine_set.go]: whoops, yeah

File: cmd/clusterctl/test/e2e/go.mod [hunk 2]
```diff
@@ -6,13 +6,13 @@ require (
 	github.com/onsi/ginkgo v1.11.0
 	github.com/onsi/gomega v1.8.1
 	github.com/pkg/errors v0.9.0
-	k8s.io/api v0.0.0-20191121015604-11707872ac1c
```
Comment (chuckha) [inline cmd/clusterctl/test/e2e/go.mod]: @fabriziopandini @Arvinderpal 

This is breaking the build so I am adding it here.

**Comment by chuckha** [comment]
This is fixed and ready for 👀

**Pull request merged**